### PR TITLE
Client-side Purging

### PR DESCRIPTION
**Pull request by SCdF**
For your consideration.

I spent a couple of hours today knocking out the basics of purge on the phone:
 - Runs every N days on refresh (in my hard-coded configuration I have it at 0 so it runs on every refresh)
 - Runs over all reports for all contacts and executes the configured function
 - Converts the result (guarding against bad ids) into minimally deleted files and writes them into PouchDB
 - Doesn't sync those deletes upwards
 - ~Initiates a compaction if anything was purged~ Does nothing because we auto-compact

I have tested it and it seems to nominally work.

Still to do:
 - Write tests!
 - Actually pull config from config
 - Document that config's existence, write a TL / CHT guide on how to use the feature
 - Performance test to make sure running like this on startup doesn't destroy a typically full phone in an obviously bad way. If it does, consider pausing between sets for n seconds to reduce constant cpu usage
 - Consider exposing this on the about page so users can manually run purge.
 - Consider putting this behind a "persisted queue", so if a user gets halfway through and refreshes we pick up where we left off.
 - Consider view warming. Doesn't really seem worth it on PouchDB, idk. Which would you pick etc.

More notes:
 - If at some point we want purge rules to only apply to certain people we could allow for the function to return a map of id to role, e.g. `['id1', 'id2']` -> `{id1: true, id2: ['role1', 'role2']}`
 - If we decide that we don't want to ship without it running on the server, while it was good practice for thinking about the problem and not much time investment I'd probably have to throw the vast majority of it away.

Interested in your thoughts.

medic/medic-webapp#5048

### CODE REVIEW

**Comment by SCdF** [comment]
Oh yeah, please don't merge 🙃

**Comment by garethbowen** [comment]
Looking good, nice work!

A few thoughts...

1. I think it's critical we never purge anything that's not yet replicated. I'm not sure what the best way to achieve this is. You could check the changes feed for the last updated date for this doc and compare that to the last sync time. Alternatively you could listen for dbsync to complete and then prompt Purger to check.
2. Seeing as you're doing this on refresh consider doing it in the bootstrap layer so the UI is blocked (nice spinner, "purging docs" message, etc). This would mean the whole process can run faster because angular wouldn't be constantly updating the UI as docs are purged.
3. We set auto compaction on in PouchDB but it would be worth considering an explicit compaction after deletion. This is a slow process but IIRC until compaction you don't see improvements in performance or disk space and I don't completely trust auto compaction to keep working. 
4. If the doc is updated on the server it'll conflict on replication. I'm guessing that'll be silently ignored which is probably the right thing to do in this case, but we could consider bringing the doc back?

**Comment by SCdF** [comment]
@garethbowen,

1. This is pretty hard, and is made redundant by running it on the server, which is another great reason to migrate to that at some point.
2. This is bad because it blocks, but good because it makes the flow cleaner and more obvious (things don't randomly delete below you), and also removes any stability concerns we have about our angular code handling deletes. This is *great* because it also half-solves replicating super large users: we'll still download all their data but we will never boot angular or even pouchdb views (I can rewrite to not use views) against the large data set.
3. OK can look into that
4. The result this is non-deterministic, but in the style of eventual consistency even if it appears again the next purge will remove it, so I figured this was OK.

So:
 - I think there are more pros than cons to moving this code to run between PouchDB bootstrap and angular bootstrap, so I'll move it there
 - I'll change it to not use views at all, and just use all docs, which sounds bad but really 99% of a user's data is going to be a contact or a report anyway, so there is little wastage
 - I will look into fixing 1.

**Comment by garethbowen** [comment]
Looking good! Let's chat sometime about the server/client tradeoffs so we can move forward here.

**Comment by alxndrsn** [comment]
> It's still not clear to me a safe way of doing this, especially since we're now doing this before angular boots.

Can you compare last-replication-time vs `doc.reported_date`?

**Comment by SCdF** [comment]
@alxndrsn unless I've missed it we don't store last-replication-time anywhere, it's only available as a live piece of data while angular is running.

We could change that so it writes it somewhere (eg to a local doc we do have access to), and then we use that on boot, presuming if that doesn't exist then we are also allowed to purge because it means no real replication has happened yet.

However, it's another write, which we're performing on every replication, to solve a check that happens once a month that probably shouldn't need to happen given good rules, and would not be reverted (since we don't need it for anything else) once we moved to server-side.

IDK, if people think it's worth it then that's cool and can be done, I personally just didn't have strong interest in it when I was knocking this together.

**Comment by garethbowen** [comment]
It's possible doc updates will still be purged before they're replicated because the code only checks the `reported_date`. Unfortunately we don't record when a doc was updated so it's not possible currently to check that.

One solution would be to move the code back into the angular app (sorry!) and trigger purging only once replication is completely caught up, but I think the faster execution in bootstrap still wins.

The other solution is running server side but we can do that in the next iteration.

This approach is good enough for now, do it!

**Comment by SCdF** [comment]
Demo: https://www.youtube.com/watch?v=laDvd0ja2ac

**Comment by SCdF** [comment]
@garethbowen THE TESTS PASS.

I am happy for this to be merged at this stage.

I'm currently writing documentation in `medic-docs`, which will be a separate PR in the next few hours. Feel free to wait for that or not, depending on your schedule.

**Comment by SCdF** [comment]
NB: I got rid of blocking purges for documents younger than 60 days, because I managed to check for upward replication.

We could add this back though, if we want to be even more careful.

Doing it for the last 30 days might be a good idea actually:
 - Means (if people are using targets properly) that targets can't be broken
 - Stops overzealous configurers confusing users by purging stuff super fast

File: webapp/src/js/bootstrapper/translator.js [hunk 1]
```diff
@@ -8,6 +8,13 @@ var TRANSLATIONS = {
       return 'Fetching info (' + x + ' docs)…';
     },
     LOAD_APP: 'Loading app…',
+    PURGE_INIT: 'Checking data…',
+    PURGE_INFO: function(progress) {
+      const n = progress.purged;
+      const percent = Math.floor((progress.processed / progress.total) * 100);
+      return `Cleaned ${n} documents (${percent}% complete)…`;
```
Comment (garethbowen) [inline webapp/src/js/bootstrapper/translator.js]: To make translation easier, perhaps do the calculation first, and call the translator with `count` and `percent`. Otherwise each language will have to do this calculation...

File: webapp/src/js/bootstrapper/translator.js [hunk 2]
```diff
@@ -8,6 +8,13 @@ var TRANSLATIONS = {
...
+    },
+    PURGE_AFTER: 'Optimising…',
```
Comment (garethbowen) [inline webapp/src/js/bootstrapper/translator.js]: Obviously these two keys need to be translated into all the languages we support. It might be difficult to get this done on short notice so feel free to raise an issue blocking the 3.4.0 release to get the translations for each language.

File: webapp/src/js/services/db-sync.js [hunk 3]
```diff
@@ -129,6 +141,11 @@ angular
 
             syncIsRecent = true;
             sendUpdate({ aggregateReplicationStatus: 'not_required' });
+
+            return DB().info()
+            .then(dbInfo => {
```
Comment (garethbowen) [inline webapp/src/js/services/db-sync.js]: This is a minor thing but it keeps happening... the rest of the codebase is formatted so that this line is indented two spaces more. eg:

```js
return DB().info()
  .then(dbInfo => {
```

If this style is something you feel strongly about then we should discuss it and update the style guide and start switching the code over to this style.

Personally I find your way hard to read because it looks like a separate statement, rather than something that's working on the return of the first line.

File: webapp/src/js/services/rules-engine.js [hunk 4]
```diff
@@ -54,7 +54,7 @@ var nools = require('nools'),
         return doc && (
           doc.patient_id ||
           doc.place_id ||
-          (doc.fields && (doc.fields.patient_id || doc.fields.place_id))
+          (doc.fields && (doc.fields.patient_id || doc.fields.place_id || doc.fields.patient_uuid))
```
Comment (garethbowen) [inline webapp/src/js/services/rules-engine.js]: Should this be a separate commit? Maybe it should even have its own issue?

File: webapp/tests/karma/unit/services/db-sync.js [hunk 5]
```diff
@@ -57,10 +61,14 @@ describe('DBSync service', () => {
...
 
   afterEach(() => {
-    KarmaUtils.restore(to, from, query, allDocs, isOnlineOnly, userCtx, sync, Auth);
+    KarmaUtils.restore(to, from, query, allDocs, info, isOnlineOnly, userCtx, sync, Auth);
   });
 
   describe('sync', () => {
+    // beforeEach(() => {
+    //   info.returns(Q.resolve({update_seq: -999}));
+    // });
```
Comment (garethbowen) [inline webapp/tests/karma/unit/services/db-sync.js]: Remove?

File: webapp/src/js/bootstrapper/index.js [hunk 6]
```diff
@@ -136,47 +138,54 @@
...
+      .then(() => purger(localDb, initialReplicationNeeded)
+        .on('start', () => setUiStatus('PURGE_INIT'))
+        .on('progress', function(progress) {
+          setUiStatus('PURGE_INFO', progress);
+        })
+        .on('optimise', () => setUiStatus('PURGE_AFTER')))
+      .then(function() {
+        // replication complete
+        setUiStatus('STARTING_APP');
+      })
```
Comment (garethbowen) [inline webapp/src/js/bootstrapper/index.js]: Ok this may be a blocker... if an admin creates a purge function that doesn't compile then the app will fail to start, right? If the app fails to start then it'll never replicate any settings updates, so fixing the function won't make it to the device, which means the phone is bricked.

Can we handle it more gracefully? eg: log the error and start up anyway and try the purge again next time?

File: webapp/src/js/bootstrapper/purger.js [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+
+    if (!highestSyncSeq) {
+      return Promise.resolve(false);
+    }
+
+    return DB.info()
+      .then(info => {
+        // I can't imagine how this would happen, but <= is arguably more
+        // correct than ===
+        if (info.update_seq <= highestSyncSeq) {
```
[the remark this answers is not in the document]
Comment (SCdF) [reply webapp/src/js/bootstrapper/purger.js]: I've wrapped it in `parseInt`. Was it confusing because you presumed it was a string?

File: webapp/src/js/bootstrapper/purger.js [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+          return true;
+        } else {
+          return false;
```
[the remark this answers is not in the document]
Comment (SCdF) [reply webapp/src/js/bootstrapper/purger.js]: Good idea, I presume I used to have more stuff going on here, changed

File: webapp/src/js/services/db-sync.js (hunk 3 above)
Comment (SCdF) [reply webapp/src/js/services/db-sync.js]: Changed

File: webapp/src/js/services/rules-engine.js (hunk 4 above)
Comment (SCdF) [reply webapp/src/js/services/rules-engine.js]: Yeah it's a good question. I like the idea of pulling it out into its own PR, because then when we merge it gets a separate commit message, so I'll do that (it'll have to be based on this PR, but that should be OK once we do the merge dance).

File: webapp/tests/karma/unit/services/db-sync.js (hunk 5 above)
Comment (SCdF) [reply webapp/tests/karma/unit/services/db-sync.js]: Whoops!

File: webapp/src/js/bootstrapper/index.js (hunk 6 above)
Comment (SCdF) [reply webapp/src/js/bootstrapper/index.js]: So this was intentional on my part, but I have been wrong before! We should have philosophical conversations once we're back from the holidays.

I just added a piece of code I planned to do next year around code execution, as it makes my intentions clearer. The situation is now:
 - If there is a parsing error the app doesn't start
 - If there is a execution error the app logs and error[1] and moves on to the next set

The reasoning behind this is:
 - We want to reduce the amount of incorrect / broken configuration in production in general
 - And here especially, as in this instance it would mean people think they are purging when they are not
 - On the one hand we shouldn't stop people from using the app
 - And on the other hand errors just logged to the console are highly likely to be ignored

So:
 - If there is a bad config we can **always** detect we should hard error so it will definitely be caught in testing. We have this philosophy in other places, such as api startup forcing you to have CouchDB configured correctly before it starts.
 - If there is bad config we can only sometimes detect (eg function execution based on parameters etc) we shouldn't hard error, because we aren't confident it will always happen, and so there is a much greater chance it will exist in production.

One addendum here that goes against my reasoning a little is that a hard error could occur because of JS feature support, which may pass if they use their local Chrome and do not also test against the phone app (which is locked to the right version). It is probably conceivable (unfortunately) that people would think to test locally, but not on the device.

[1] I am considering taking the execution errors and putting them somewhere feedback can find them on startup and send out. Maybe that's too extra though?

Comment (garethbowen) [reply webapp/src/js/bootstrapper/index.js]: Hard failures are good but only if the phone can recover when the error is fixed on the server. Unless I've missed something if the function fails to compile then potentially every phone will have to be brought back to base and wiped and everything downloaded again. This failure is _too_ hard.

I'm happy with the app not starting up, but only if it can be fixed remotely without IT support. I can't think of any easy way to do that. Maybe starting the app with a parameter which runs replication but obscures the UI. Then when the settings are changed the dialog will show, the user can refresh, and be up and running again.

**Comment by SCdF** [comment]
@garethbowen OK check out https://github.com/medic/medic-webapp/pull/5075/commits/3f7a58cb939ac9e8a6c1fc74beaaa69c7e7804bf

We no longer ever hard fail for anything.

Instead, if the purge function cannot be compiled, or a purge set fails for some reason, we generate feedback errors (up to 10 to avoid spam), which are then generated into real feedback documents on successful Angular boot.

Additionally, in testing I decided that it makes sense that purge should re-run if you change the purge function, so I added that as well. It's going to be an interesting decision when moving from client to server, but that's a conversation for another time.
